docs/uenv-compilation-spack.md

```diff
@@ -0,0 +1,163 @@
+
+!!! note
+    It is good practice to have a single root [spec] in an environment, and to define constraint on all packages or specific dependencies in the `packages:` field.
+
+### Building
+
+After defining the environment above, we can concretize it:
+
+```bash
+spack -e SPACK_ENV_FOLDER concretize -f
+```
+
+The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked ad `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly speed up compilation, since [Spack] will have to build only a small subset of packages.
+
+You can finally build everything in the concretized environment:
+
+```bash
+spack -e SPACK_ENV_FOLDER install
+```
+
+## Developing with Spack
+
+In addition to wanting to build a different configuration of a package as described above, you might want to build your own software from source. Let's assume we want to develop [CP2K] with the COSMA library as a dependency:
+
+```
+cp2k@2024.1 +cuda cuda_arch=80 +cosma ^cosma +gpu_direct
+```
+
+We assume again that COSMA is not present in the provided uenv.
+
+### Spack Environment and Building Dependencies
+
+As described above, you can define a [Spack environment] describing the version of the package you want to build and the contraint on the dependencies. After concretizing the environment with
+
+```bash
+spack -e SPACK_ENV_FOLDER concretize -f
+```
+
+we can tell [Spack] to only install the dependencies, since we want to build the root [spec] manually:
+
+```bash
+spack -e SPACK_ENV_FOLDER install --only=dependencies
+```
+
+### Building the root spec manually
+
+Finally, we are ready to build the root [spec] manually. With [Spack] you can get a shell within the build environment as follows:
+
+```bash
+spack -e SPACK_ENV_FOLDER build-env cp2k -- bash
+```
+
+where `cp2k` denotes the root [spec]. Since there is only one such [spec], there is no need to explicitly write out the version nor the variants.
+
+Within the build environment, the software can be built using the provided build system. CP2K uses CMake, therefore we can simply do the following:
+
+```bash
+# In CP2K repository root folder
+
+mkdir build && cd build
+
+cmake \
+    -GNinja
+    -DCP2K_USE_COSMA=ON \
+    -DCP2K_USE_ACCEL=CUDA \
+    -DCP2K_WITH_GPU=A100 \
+    ..
+
+ninja -j 32
+```
+
+## Known Limitations
+
+!!! warning
+    Swapping the upstream [Spack] instance by loading different uenvs might lead to a corruption of the [Spack] database. If this happens, you can uninstall everything from your local [Spack] instance with `spack uninstall --all` and clean up with `spack clean --all`.
```

Comment:
```suggestion
    Swapping the upstream [Spack] instance by loading different uenvs might lead to surprising inconsistencies in the [Spack] database. If this happens, you can uninstall everything from your local [Spack] instance with `spack uninstall --all` and clean up with `spack clean --all`.
```
maybe? This warning should be scary, but "corruption" may be a bit too strong (since `spack reindex` can make it somewhat usable again with a new uenv).

Reply:
I'm wondering if we should add `config:install_tree:root:$spack/<hash of uenv recipe>/opt/spack` to `/user-environment/config.yaml`? Then each uenv would have it's own spack install tree. Or if it's better to let the user manage it consciously by creating different spack clones on scratch (if needed at all).